Propose ADR 027: Deterministic Protobuf Serialization (#6979)
cosmos/cosmos-sdk: docs/architecture/adr-026-protobuf-regencode.md

to be used by other ADRs.

### Scope

Regencode is a protobuf3 serializer. The output is a valid protobuf
serialization, such that every protobuf parser can parse it.

No maps are supported in version 1 due to the complexity of defining a
derterministic serialization. This might change in future. Implementations must
reject documents containing maps as invalid input.

### Serialization rules

The serialization is based on the
[protobuf 3 encoding](https://developers.google.com/protocol-buffers/docs/encoding)
with the following additions:

1. Fields must be serialized in ascending order
2. Extra fields or any extra data must not be added
3. [Default values](https://developers.google.com/protocol-buffers/docs/proto3#default)
   must be omitted

While rule number 1. and 2. should be pretty straight forward and describe the
default behaviour of all protobuf encoders the author is aware of, the 3rd rule
is more interesting. Protobuf 3 has no notion of optional values and thus cannot
differentiate between null/unset and empty fields. At serialization level
however, it is possible to set the fields with an empty value or omitting them

> entirely. This is a significat difference to e.g. JSON where a property can be

Reviewer: ```suggestion
entirely. This is a significant difference to e.g. JSON where a property can be
```

empty (`""`, `0`), `null` or undefined, leading to 3 different documents.

> Omitting empty fields is a valid option because the parser must assign the
> default value to fields missing in the serialization<sup>2</sup>. Requiring to
> serialize them would be equally valid. No good reason is known to the author for
> preferring one of those options over the other.

Reviewer: ```suggestion
Omitting empty fields is the logical choice for a canonical protobuf serialization because it allows for some amount of forward compatibility. That means that users of newer versions of a protobuf schema will produce the same serialization as users of older versions if they do not set newer fields. Effectively omitting newer fields causes the newer clients to fall back to an older version of the protocol. This would not be possible if empty fields were serialized by default.
```

### Implementation

> There are three main implementation strategies, orderd from the least to the

Reviewer: ```suggestion
There are three main implementation strategies, ordered from the least to the
```

most custom development:

- **Use a protobuf serializer that follows the above rules by default.** E.g.
  [gogoproto](https://pkg.go.dev/github.com/gogo/protobuf/gogoproto) is known to

>   be compliant by default. It might also be an option to configure an existing

Reviewer: ```suggestion
  be compliant by in most cases, but not when certain annotations such as `nullable = false` are used. It might also be an option to configure an existing
```

>   serializer accoringly.

Reviewer: ``
```suggestion
  serializer accordingly.
```

- **Normalize default values before encoding them.** If your serializer follows
  rule 1. and 2. and allows you to explicitely unset fields for serialization,
  you can normalize empty values to unset. This can be done when working with
  [protobuf.js](https://www.npmjs.com/package/protobufjs):

  ```js
  const bytes = SignDoc.encode({
    bodyBytes: body.length > 0 ? body : null, // normalize empty bytes to unset
    authInfoBytes: authInfo.length > 0 ? authInfo : null, // normalize empty bytes to unset
    chainId: chainId || null, // normalize "" to unset
    accountNumber: accountNumber || null, // normalize 0 to unset
    accountSequence: accountSequence || null, // normalize 0 to unset
  }).finish();
  ```

> - **Use a hand-written serializer for the types you need.** If non of the above

Reviewer: ```suggestion
- **Use a hand-written serializer for the types you need.** If none of the above
```
